File: frontend/src/api/schemas.ts
import { z } from 'zod'

/**
 * Zod schemas for REST responses — the single source of truth for API shapes. Types are derived via
 * `z.infer`, and the query hooks `.parse()` responses at the boundary, so backend drift throws loudly
 * here instead of surfacing as a silent runtime bug three components deep. Fields the UI can tolerate
 * missing get `.default(...)` so a thin/legacy payload degrades rather than rejects.
 */

export const HuntSummarySchema = z.object({
  hunt_id: z.string(),
  state: z.string(),
  source: z.string().default(''),
  title: z.string().default(''),
  // Backend sends null until a plan is approved; `.default` only fills undefined, so it must be
  // nullable or the whole list parse throws and the Past Hunts sidebar silently shows empty.
  boundary_usd: z.number().nullable().default(null),
  project_id: z.string().nullable().default(null),
  created_at: z.string().default(''),
  cost_usd: z.number().default(0),
})
export type HuntSummary = z.infer<typeof HuntSummarySchema>

export const HuntListResponseSchema = z.object({
  hunts: z.array(HuntSummarySchema),
  next_cursor: z.string().nullable(),
})
export type HuntListResponse = z.infer<typeof HuntListResponseSchema>

export const HuntSnapshotSchema = z.object({
  hunt_id: z.string(),
  state: z.string(),
  last_seq: z.number(),
  task: z.string().default(''),
  strategy: z.string().default('orchestrate'),
  project_id: z.string().nullable().default(null),
  created_at: z.string().nullable().default(null),
  updated_at: z.string().nullable().default(null),
})
export type HuntSnapshot = z.infer<typeof HuntSnapshotSchema>

export const ArtifactKindSchema = z.enum(['md', 'html', 'pdf', 'docx', 'xlsx', 'pptx', 'png'])
export type ArtifactKind = z.infer<typeof ArtifactKindSchema>

export const ArtifactMetaSchema = z.object({ artifact_id: z.string(), kind: ArtifactKindSchema })
export type ArtifactMeta = z.infer<typeof ArtifactMetaSchema>

export const BriefSourceSchema = z.object({
  title: z.string().default(''),
  url: z.string().default(''),
  by: z.string().default(''),
  verified: z.boolean().default(false),
  timestamp: z.string().optional(),
})
export type BriefSource = z.infer<typeof BriefSourceSchema>

export const BriefBlockSchema = z.object({
  text: z.string().default(''),
  source_ids: z.array(z.number()).default([]),
})
export type BriefBlock = z.infer<typeof BriefBlockSchema>

export const BriefContentSchema = z.object({
  text: z.string().default(''),
  blocks: z.array(BriefBlockSchema).default([]),
  claims: z.array(z.string()).default([]),
  sources: z.array(BriefSourceSchema).default([]),
  span_map_ref: z.string().nullable().default(null),
  no_sources: z.boolean().default(false),
  refined: z.boolean().optional(),
})
export type BriefContent = z.infer<typeof BriefContentSchema>

export const BriefSchema = z.object({
  artifact_id: z.string(),
  hunt_id: z.string(),
  kind: z.string(),
  produced_by: z.string().nullable().default(null),
  content: BriefContentSchema,
})
export type Brief = z.infer<typeof BriefSchema>

export const SharedSchema = z.object({
  title: z.string().default(''),
  content: BriefContentSchema.nullable(),
})
export type Shared = z.infer<typeof SharedSchema>

export const ScorecardSideSchema = z.object({
  quality: z.number().default(0),
  citations: z.number().default(0),
  cost_usd: z.number().default(0),
  time_s: z.number().default(0),
  sources: z.number().default(0),
})
export type ScorecardSide = z.infer<typeof ScorecardSideSchema>

export const ScorecardSchema = z.object({
  lone_wolf: ScorecardSideSchema,
  pack: ScorecardSideSchema,
})
export type Scorecard = z.infer<typeof ScorecardSchema>

/** GET /hunts/:id/receipts — per-claim provenance for the delivered brief. */
export const ReceiptSourceSchema = z.object({
  n: z.number(),
  title: z.string().default(''),
  url: z.string().default(''),
  by: z.string().default(''),
  verified: z.boolean().default(false),
  library: z.boolean().default(false),
})
export const ReceiptClaimSchema = z.object({
  text: z.string(),
  // an unknown future status degrades to the neutral middle instead of rejecting the payload
  status: z.enum(['verified', 'cited', 'unsourced', 'challenged_kept']).catch('cited'),
  sources: z.array(ReceiptSourceSchema).default([]),
  challenge: z.object({ problem: z.string().default('') }).nullable().default(null),
})
export const ReceiptsSchema = z.object({
  hunt_id: z.string().default(''),
  critique_ran: z.boolean().default(false),
  review_note: z.string().default(''),
  claims: z.array(ReceiptClaimSchema).default([]),
  dropped: z.array(z.object({ text: z.string(), problem: z.string().default('') })).default([]),
  standoff: z
    .object({
      challenger: z.string().default('sentinel'),
      defendant: z.string().default('tracker'),
      outcome: z.string().default('unresolved'),
      rationale: z.string().default(''),
    })
    .nullable()
    .default(null),
  wolves: z
    .record(z.string(), z.object({ sources: z.number().default(0), verified: z.number().default(0) }))
    .default({}),
  documents: z
    .array(z.object({ doc_id: z.string(), title: z.string().default(''), cited_by_claims: z.number().default(0) }))
    .default([]),
  totals: z.record(z.string(), z.number()).default({}),
})
export type ReceiptClaim = z.infer<typeof ReceiptClaimSchema>
export type Receipts = z.infer<typeof ReceiptsSchema>

/** GET /share/:token/tracks — the public Flight Recorder: a shared hunt's full redacted event
 *  log. Events stay loosely typed here; the replay engine re-parses each one against
 *  HuntEventSchema and skips anything it doesn't recognize (same policy as the live stream). */
export const SharedTracksSchema = z.object({
  title: z.string().default('A Pack hunt'),
  events: z.array(z.record(z.string(), z.unknown())).default([]),
  redacted: z.boolean().default(true),
})
export type SharedTracks = z.infer<typeof SharedTracksSchema>

/** POST /hunts/:id/rehearse — the Shadow Hunt cost/time estimate for a team + strategy + depth.
 *  Pure and instant server-side (no model calls) — safe to re-query as the user reshapes the pack. */
export const RehearseSchema = z.object({
  est_cost_usd: z.number(),
  est_time_s: z.number(),
  calls: z.number(),
  scouts: z.number(),
  warnings: z.array(z.string()).default([]),
})
export type Rehearse = z.infer<typeof RehearseSchema>

export const SpendHuntSchema = z.object({
  hunt_id: z.string(),
  title: z.string().default(''),
  cost_usd: z.number().default(0),
})
export const SpendSummarySchema = z.object({
  total_usd: z.number().default(0),
  hunts: z.array(SpendHuntSchema).default([]),
})
export type SpendHunt = z.infer<typeof SpendHuntSchema>
export type SpendSummary = z.infer<typeof SpendSummarySchema>

File: backend/app/schemas.py
"""All Pydantic request + response models for the Pack API.

Request models validate inbound bodies. Response models document and enforce what each route
returns — every route in app/routers/ carries response_model= pointing here.
"""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

# ---------------------------------------------------------------------------
# Shared enums (reject junk with 422 before it reaches the engine)
# ---------------------------------------------------------------------------

Strategy = Literal["orchestrate", "deep_dive", "critique"]
Source = Literal["typed", "spoken", "dropped"]
Mode = Literal["wild", "on_signal", "on_command"]
Depth = Literal["brief", "standard", "deep"]
InputKind = Literal["text", "pdf", "csv", "md", "url", "image", "audio", "video"]

_MAX_TASK = 10_000
_MAX_INPUT = 200_000

# ---------------------------------------------------------------------------
# Request models
# ---------------------------------------------------------------------------


class CreateHunt(BaseModel):
    input: str | None = Field(None, max_length=_MAX_TASK, description="The task.")
    instinct_id: str | None = Field(None, max_length=120)
    source: Source = "typed"
    strategy: Strategy | None = Field(None, description="Research strategy.")
    team: list[dict] | None = Field(None, max_length=16)


class ApprovePlan(BaseModel):
    mode: Mode
    boundary_usd: float = Field(..., ge=0, le=1000, description="Dollar Boundary.")
    edits: dict | None = None
    depth: Depth | None = Field(
        None, description="v3: user override of the brief's depth; None keeps Beta's."
    )


class ResolveHold(BaseModel):
    resolution: str = Field(..., max_length=2000)
    edited_text: str | None = Field(None, max_length=_MAX_INPUT)


class ResumeHunt(BaseModel):
    boundary_usd: float = Field(..., ge=0, le=1000)


class SaveInstinct(BaseModel):
    label: str = Field(..., min_length=1, max_length=200)
    spec: dict = Field(default_factory=dict)


class InstinctPatch(BaseModel):
    label: str | None = Field(None, min_length=1, max_length=200)
    spec: dict | None = None


class HuntPatch(BaseModel):
    title: str | None = Field(None, max_length=200)
    archived: bool | None = None
    project_id: str | None = None  # presence-checked; null explicitly unassigns


class MessageIn(BaseModel):
    role: Literal["user", "alpha"]
    content: str = Field(..., max_length=_MAX_INPUT)


class ProjectIn(BaseModel):
    label: str = Field(..., min_length=1, max_length=200)
    instructions: str | None = Field(None, max_length=10_000)


class ProjectPatch(BaseModel):
    label: str | None = Field(None, max_length=200)
    instructions: str | None = Field(None, max_length=10_000)


class AddInput(BaseModel):
    text: str = Field(..., max_length=_MAX_INPUT, description="Text to fold into the hunt.")
    kind: InputKind = "text"


class TeamEntry(BaseModel):
    """One formation slot. `role` is free-form (rehearse warns on unknown roles rather than
    rejecting); `count` is validated/coerced here so a malformed value (e.g. a non-numeric string)
    422s at the boundary instead of crashing rehearse() with an uncaught ValueError."""

    model_config = ConfigDict(extra="ignore")
    role: str = ""
    count: int = Field(default=0, ge=0, le=50)


class RehearseBody(BaseModel):
    team: list[TeamEntry] | None = None
    strategy: str | None = None
    depth: Depth | None = None


class FeedbackBody(BaseModel):
    turn_index: int
    vote: str = Field(..., pattern="^(up|down)$")


class RefineBody(BaseModel):
    instruction: str = Field("", max_length=2000, description="How to re-angle/tighten the brief.")


class AskAlpha(BaseModel):
    question: str | None = Field(None, max_length=_MAX_TASK)
    messages: list[dict] = Field(default_factory=list, max_length=200)


class IntakeBody(BaseModel):
    messages: list[dict] = Field(default_factory=list, max_length=200)
    # The hunt this conversation is attached to, if one already exists. Lets the front-door gate see
    # the live hunt state (running/delivered) so it stops re-asking scoping questions after a hunt has
    # started, and doesn't relaunch after one delivered. None during a genuinely fresh intake.
    hunt_id: str | None = Field(None, max_length=64)


# ---------------------------------------------------------------------------
# Generic response shapes
# ---------------------------------------------------------------------------


class OkResponse(BaseModel):
    ok: bool


class ClearedResponse(BaseModel):
    cleared: bool


# ---------------------------------------------------------------------------
# Hunt response models
# ---------------------------------------------------------------------------


class HuntCreated(BaseModel):
    hunt_id: str
    state: str


class HuntSnapshot(BaseModel):
    hunt_id: str
    state: str
    last_seq: int
    task: str = ""
    strategy: str = "orchestrate"
    project_id: str | None = None
    created_at: str | None = None
    updated_at: str | None = None


class HuntSummary(BaseModel):
    model_config = ConfigDict(extra="allow")
    hunt_id: str
    state: str
    title: str = ""
    cost_usd: float = 0.0


class HuntListResponse(BaseModel):
    hunts: list[HuntSummary]
    next_cursor: str | None = None


class HuntPatchResponse(BaseModel):
    hunt_id: str
    ok: bool


class HuntDeleteResponse(BaseModel):
    hunt_id: str
    deleted: bool


class CommandAccepted(BaseModel):
    hunt_id: str
    accepted: bool


class IntakeReply(BaseModel):
    reply: str
    ready: bool
    brief: str


class AskReply(BaseModel):
    reply: str
    # What Alpha DID with this turn, so the frontend can react (refresh the brief on a refine, track a
    # spawned follow-up hunt, etc.). "answer" = a plain reply (the legacy behaviour). "refined" = the
    # brief was re-drafted (refresh the reward). "subhunt" = a scoped follow-up hunt was launched to
    # extend the brief (its id is on `hunt_id`). "new_hunt" = a fresh hunt (its id on `hunt_id`).
    action: str = "answer"
    hunt_id: str | None = None


class MessageItem(BaseModel):
    role: str
    text: str


class MessagesResponse(BaseModel):
    messages: list[MessageItem]


class ShareResponse(BaseModel):
    token: str


class SharedResponse(BaseModel):
    model_config = ConfigDict(extra="allow")
    title: str
    content: dict[str, Any] | None = None


class FeedbackResponse(BaseModel):
    votes: list[dict[str, Any]]
    up: int
    down: int


class RehearseResponse(BaseModel):
    est_cost_usd: float
    est_time_s: float
    calls: int
    scouts: int
    warnings: list[str]


class ArtifactMeta(BaseModel):
    artifact_id: str
    kind: str


class ArtifactsListResponse(BaseModel):
    artifacts: list[ArtifactMeta]


class ArtifactResponse(BaseModel):
    model_config = ConfigDict(extra="allow")
    artifact_id: str
    hunt_id: str
    kind: str
    produced_by: str | None = None
    content: dict[str, Any] | None = None


class RefineResponse(BaseModel):
    hunt_id: str
    artifact_id: str
    accepted: bool


class ScorecardResponse(BaseModel):
    hunt_id: str
    scorecard: dict[str, Any]


class TracksResponse(BaseModel):
    hunt_id: str
    events: list[dict[str, Any]]
    redacted: bool


class SharedTracksResponse(BaseModel):
    """The public Flight Recorder: a shared hunt's full redacted event log, keyed by the share
    token (never a hunt id) so the link's scope is exactly one hunt's replay."""

    title: str
    events: list[dict[str, Any]]
    redacted: bool


class ReceiptsResponse(BaseModel):
    """The Receipts — per-claim provenance for a delivered brief: each claim's sources (with the
    wolf that found each and whether the page was read), the Sentinel's challenges and their
    outcomes, the claims that were dropped, and your-documents coverage."""

    hunt_id: str
    critique_ran: bool
    review_note: str = ""  # why verification didn't complete, when critique_ran is False
    claims: list[dict[str, Any]]
    dropped: list[dict[str, Any]]
    standoff: dict[str, Any] | None = None
    wolves: dict[str, dict[str, int]]
    documents: list[dict[str, Any]]
    totals: dict[str, int]


# ---------------------------------------------------------------------------
# Project response models
# ---------------------------------------------------------------------------


class ProjectItem(BaseModel):
    model_config = ConfigDict(extra="allow")
    project_id: str
    label: str


class ProjectsListResponse(BaseModel):
    projects: list[ProjectItem]


class ProjectResponse(BaseModel):
    model_config = ConfigDict(extra="allow")
    project_id: str
    label: str


class ProjectCreatedResponse(BaseModel):
    project_id: str
    label: str


class ProjectPatchResponse(BaseModel):
    project_id: str
    ok: bool


class ProjectDeleteResponse(BaseModel):
    project_id: str
    deleted: bool


# ---------------------------------------------------------------------------
# Instinct response models
# ---------------------------------------------------------------------------


class InstinctItem(BaseModel):
    instinct_id: str
    label: str
    spec: dict[str, Any] = Field(default_factory=dict)


class InstinctsListResponse(BaseModel):
    instincts: list[InstinctItem]


class InstinctCreatedResponse(BaseModel):
    instinct_id: str
    accepted: bool


class InstinctPatchResponse(BaseModel):
    instinct_id: str
    ok: bool


class InstinctDeleteResponse(BaseModel):
    instinct_id: str
    deleted: bool


# ---------------------------------------------------------------------------
# Document response models
# ---------------------------------------------------------------------------


class DocMeta(BaseModel):
    id: int
    name: str
    kind: str
    chars: int


class DocumentsListResponse(BaseModel):
    documents: list[DocMeta]


class DocumentResponse(BaseModel):
    model_config = ConfigDict(extra="allow")
    id: int
    name: str
    kind: str
    chars: int


class DocumentDetailResponse(BaseModel):
    model_config = ConfigDict(extra="allow")
    id: int
    name: str
    kind: str
    chars: int
    text: str


class DocumentDeleteResponse(BaseModel):
    id: int
    deleted: bool


# ---------------------------------------------------------------------------
# Memory / Spend response models
# ---------------------------------------------------------------------------


class MemoryItem(BaseModel):
    # v6: id + status make each lesson addressable — editable, vetoable, and citable (memory://id).
    id: int
    text: str
    # The lesson type the Elder assigned — what-worked / what-failed / preference / topic-insight,
    # or the legacy "takeaway" for older/untyped rows — so the UI can group and label what it shows.
    kind: str = "takeaway"
    hunt_id: str | None = None
    status: str = "active"


class MemoryResponse(BaseModel):
    memory: list[MemoryItem]


class MemoryPatch(BaseModel):
    """Edit a lesson: rewrite its text and/or flip its lifecycle (active ↔ archived). Archived =
    vetoed — kept for the record, never recalled into a hunt again."""

    text: str | None = None
    status: str | None = None


class MemoryPatchResponse(BaseModel):
    ok: bool


class MemoryDeleteResponse(BaseModel):
    deleted: bool


class SpendHuntItem(BaseModel):
    model_config = ConfigDict(extra="allow")
    hunt_id: str
    cost_usd: float


class SpendResponse(BaseModel):
    total_usd: float
    hunts: list[SpendHuntItem]


# ---------------------------------------------------------------------------
# System response models
# ---------------------------------------------------------------------------


class HealthResponse(BaseModel):
    model_config = ConfigDict(extra="allow")
    status: str
    service: str


class ReadyResponse(BaseModel):
    ready: bool


class StrategyInfo(BaseModel):
    model_config = ConfigDict(extra="allow")
    name: str
    label: str
    pattern: str


class StrategiesResponse(BaseModel):
    strategies: list[StrategyInfo]
    default: str


# ---------------------------------------------------------------------------
# Parse / Transcribe response models
# ---------------------------------------------------------------------------


class ParsedDocResponse(BaseModel):
    kind: str
    text: str
    chars: int
    filename: str | None = None


class TranscriptResponse(BaseModel):
    text: str
    provider: str
    duration_s: float
